### DISCUSSION TITLE
Button in Tracker?

### DISCUSSION THREAD

**Discussion by redundancer** [oxyplot/oxyplot#1981] [Q&A] [answered]
Is it possible to add a button with OnClick/ Command functionality to a tracker? If yes, how?

I am panning to have a button in the tracker of a ScatterSeries. If you click the button, another plot will be visible with more information about the point.

Thanks in advance!

**Comment by VisualMelon** [comment]
It may be possible with some platforms, e.g. there are examples for custom trackers for WPF: https://github.com/oxyplot/oxyplot/blob/develop/Source/Examples/WPF/WpfExamples/Examples/CustomTrackerDemo/MainWindow.xaml

Note that you'll need a custom plot controller in order to keep the tracker open after clicking: https://github.com/oxyplot/oxyplot/blob/899ea5b3baf3d2010680f2e607ad5281e3f190e5/Source/Examples/WPF/SimpleDemo/CustomPlotController.cs

In general, may be best to handle clicking of the series manually without the tracker, as it sounds like you won't really be using it as a tracker any more, and that should be achievable on any platform (WinForms, for example, doesn't support custom trackers)

**Reply by redundancer** [reply]
Thank you this helped! I could configure the controller so that it does not close after releasing the left mouse button. My TrackerControlTemplate looks like this:

```
                    <oxy:PlotView.DefaultTrackerTemplate>
                        <ControlTemplate>
                            <oxy:TrackerControl Position="{Binding Position}" LineExtents="{Binding PlotModel.PlotArea}"
                                                BorderThickness="1"
                                                >
                                <oxy:TrackerControl.Background>
                                    <SolidColorBrush Color="LightBlue" />
                                </oxy:TrackerControl.Background>
                                <oxy:TrackerControl.Content>
                                    <StackPanel Orientation="Vertical">
                                        <TextBlock Text="{Binding}" Margin="7" />
                                        <StackPanel Orientation="Horizontal">
                                            <Button Command="{Binding SwitchToDvhsTabCommand}">DVHs</Button>
                                            <Button Command="{Binding SwitchToObjectivesTabCommand}">Objectives</Button>
                                        </StackPanel>
                                    </StackPanel>
                                </oxy:TrackerControl.Content>
                            </oxy:TrackerControl>
                        </ControlTemplate>
                    </oxy:PlotView.DefaultTrackerTemplate>
```
The problem is now that it cannot bind `SwitchToDvhsTabCommand` and `SwitchToObjectivesTabCommand`. They are defined in the ViewModel and I bind the view model with [Prism](https://prismlibrary.com/docs/wpf/legacy/Implementing-MVVM.html). All other bindings in this Model and ViewModel work properly. I think I am just not familiar engough with WPF. Do you know what I am doing wrong?

**Reply by VisualMelon** [reply]
It's probably that the `TrackerControl` has its own (set by OxyPlot) data-context: you'll need to qualify the binding so that you access your own data context. (e.g. think about what the `TextBlock` is binding to: I'm guessing it's not the data-context you want for the buttons)

It's been years since I did anything non-trivial in WPF, so I'm not sure of the syntax, but I think you need to set the `RelativeSource` or `ElementName` or similar in the binding string to e.g. use the data-context of the parent window/some named element.

**Reply by redundancer** [reply]
Youe are right, the problem is that it tries to find the commands inside TrackerHitResult, if I understand correctly:

![grafik](https://user-images.githubusercontent.com/81110847/223160834-6c8764d5-a9a3-4298-ab81-754f834db789.png)

It does not throw an error anymore when I use the following:
```
<Button Command="{Binding DataContext.SwitchToDvhsTabCommand,
             RelativeSource={RelativeSource AncestorType=UserControl}}">DVHs</Button>
<Button Command="{Binding DataContext.SwitchToObjectivesTabCommand, 
              RelativeSource={RelativeSource AncestorType=UserControl}}">Objectives</Button>
```
but it also does not execute any command :( However, as soo as I click on a scatter point, it runs the CanExecute commands. Very strange...

**Reply by VisualMelon** [reply]
Perhaps `CanExecute` is being bound before the data-context is changed?

**Reply by redundancer** [reply]
I don't know how to find that out. I also tried putting it into code behind, but this does also not fire...

**Reply by redundancer** [reply]
Is it possible that Buttons inside a TrackerControl cannot be clicked for some reason? I also do not get a visual response. Neither on click, nor on hover.

**Reply by VisualMelon** [reply]
You wouldn't get a visual response if the button is disabled, which I imagine it would be if it can't bind the command. Maybe bind some text or similar which will provide a visual clue if it is bound correctly.

**Reply by redundancer** [reply]
Ok. I bound the content of the buttons to a string property, exactly like the command, and the content works! This is so frustrating right now...
```
<Button Command="{Binding DataContext.SwitchToObjectivesTabCommand, ElementName=RtOptimizationUserControl}"
              Content="{Binding DataContext.ObjectivesButtonText, ElementName=RtOptimizationUserControl}"/>
```

**Reply by VisualMelon** [reply]
Looks to me like you're right: the theme does set `HitTestVisible=false` for `TrackerControl`; should be overridable in your own style

**Reply by redundancer** [reply]
I've put the `IsHitTestVisible="True"` everywhere I could, but still nothing:
```
<oxy:PlotView.DefaultTrackerTemplate>
      <ControlTemplate>
          <oxy:TrackerControl Name="OxyTrackerControl"
              Position="{Binding Position}" LineExtents="{Binding PlotModel.PlotArea}"
                              BorderThickness="1"
                              IsHitTestVisible="True">
              <oxy:TrackerControl.Background>
                  <SolidColorBrush Color="LightBlue" />
              </oxy:TrackerControl.Background>
              <oxy:TrackerControl.Content>
                  <StackPanel Orientation="Vertical" IsHitTestVisible="True">
                      <TextBlock Text="{Binding}" Margin="7"/>
                      <StackPanel Orientation="Horizontal">
                          <Button IsHitTestVisible="True"
                              Command="{Binding DataContext.SwitchToDvhsTabCommand, ElementName=RtOptimizationUserControl}">DVHs</Button>
                          <Button  IsHitTestVisible="True"
                                   Command="{Binding DataContext.SwitchToObjectivesTabCommand, ElementName=RtOptimizationUserControl}"
                                   Content="{Binding DataContext.ObjectivesButtonText, ElementName=RtOptimizationUserControl}"/>
                      </StackPanel>
                  </StackPanel>
              </oxy:TrackerControl.Content>
          </oxy:TrackerControl>
      </ControlTemplate>
  </oxy:PlotView.DefaultTrackerTemplate>
```

**Reply by redundancer** [reply]
Also tried it this way without success:
```
<oxy:PlotView
                    Model="{Binding OptimizationPlotModel, Mode=TwoWay}"
                    Controller="{Binding CustomPlotController}">
                    <oxy:PlotView.Resources>
                        <Style TargetType="oxy:TrackerControl">
                            <Setter Property="IsHitTestVisible" Value="True" />
                        </Style>
                        <Style TargetType="StackPanel">
                            <Setter Property="IsHitTestVisible" Value="True" />
                        </Style>
                        <Style TargetType="Button">
                            <Setter Property="IsHitTestVisible" Value="True" />
                        </Style>
                    </oxy:PlotView.Resources>
                    <oxy:PlotView.DefaultTrackerTemplate>
                        <ControlTemplate>
                            <oxy:TrackerControl Name="OxyTrackerControl"
                                Position="{Binding Position}" LineExtents="{Binding PlotModel.PlotArea}"
                                                BorderThickness="1"
                                                IsHitTestVisible="True">
                                <oxy:TrackerControl.Background>
                                    <SolidColorBrush Color="LightBlue" />
                                </oxy:TrackerControl.Background>
                                <oxy:TrackerControl.Content>
                                    <StackPanel Orientation="Vertical">
                                        <TextBlock Text="{Binding}" Margin="7"/>
                                        <StackPanel Orientation="Horizontal">
                                            <Button Command="{Binding DataContext.SwitchToDvhsTabCommand, ElementName=RtOptimizationUserControl}">DVHs</Button>
                                            <Button Command="{Binding DataContext.SwitchToObjectivesTabCommand, ElementName=RtOptimizationUserControl}"
                                                     Content="{Binding DataContext.ObjectivesButtonText, ElementName=RtOptimizationUserControl}"/>
                                        </StackPanel>
                                    </StackPanel>
                                </oxy:TrackerControl.Content>
                            </oxy:TrackerControl>
                        </ControlTemplate>
                    </oxy:PlotView.DefaultTrackerTemplate>
                </oxy:PlotView>
```

**Reply by redundancer** [reply]
Also not in the surrounding UserControl:
```
    <UserControl.Resources>
        <Style TargetType="oxy:TrackerControl">
            <Setter Property="IsHitTestVisible" Value="True" />
        </Style>
    </UserControl.Resources>
```

**Reply by VisualMelon** [reply]
I think the problem is that the tracker is below the invisible interaction pane that sits on top of the plot. This may make this rather more difficult than I'd thought.

Yeah, the tracker is put in the `overlays` part of the `PlotView`, which is below a transparent `mouseGrid` that which eats all user input: https://github.com/oxyplot/oxyplot/blob/196bb74bd8105bc1b5730389ec779996e57156a9/Source/OxyPlot.Wpf.Shared/PlotViewBase.cs#L178-L206

Off-hand, I can't think how to get around this...

**Reply by redundancer** [reply]
Oh no :( Thank you very much for your effort!

**Reply by VisualMelon** [reply]
The commit that added the transparent overlay is from https://github.com/oxyplot/oxyplot/issues/382

I did some tests, and moving the `overlays` in front of the `mouseGrid` didn't seem to cause any issues for me... so it may be enough to just swap them over. This of course may lead to issues with updates as described in #382, but maybe that's acceptable for your use-case (I don't understand #382, so won't try to predict what will happen). This could be done by overriding `OnApplyTemplate` on `PlotView` and swapping the last 2 children of the grid after calling the base (or I vaguely recall WPF has Z-index support within panels, but again been a while). This combined with setting `IsHitTsetVisible` on `TrackerControl` should be enough, based on some rough testing.

Obviously overloading `PlotView` is a bit of a pain: can always bodge it if necessary e.g. by hooking something that will be called after `OnApplyTemplate`.

**Reply by redundancer** [reply]
Thank you so much! I tried around a lot and it seems like removing the `mouseGrid` solves my problem. Every functionality i need still works and I can live with not recognizing click events during `InvalidatePlot()` as described in #382. I will post an answer with all details in case someone else wants to do same thing as me.

**Comment by redundancer** [comment, marked as answer]
With a lot of help from @VisualMelon I finally managed to create a tracker with working buttons inside. To summarize:
1. To avoid closing of the tracker after release of the left mouse button I created a `TrackerManipulator`:
```csharp
internal class StaysOpenTrackerManipulator : OxyPlot.TrackerManipulator
  {
      public StaysOpenTrackerManipulator(IPlotView plotView) : base(plotView)
      {
          Snap = true;
          PointsOnly = true;
      }

      public override void Started(OxyMouseEventArgs e)
      {
           /*
             * Close tracker when clicking. This avoids wrong shaping of the tracker
             * in case it's text is of different length for different points. Additionally,
             * one can close the tracker like this by clicking somewhere else.
             */
          PlotView.HideTracker();
          base.Started(e);
      }

      public override void Completed(OxyMouseEventArgs e)
      {
          // Do nothing to avoid tracker closing.
      }
  }
```
and add it to the plot controller:
```csharp
CustomPlotController.BindMouseDown(mouseButton, new DelegatePlotCommand<OxyMouseDownEventArgs>((view, controller, args) =>
                controller.AddMouseManipulator(view, new StaysOpenTrackerManipulator(view), args)));
```
2. A transparent `mouseGrid` that is on top of the tracker prevents the tracker control to recognize hits. Removing that `mouseControl` solved it for me, but it may cause problems described here #382. Hence I created my own very simple PlotView:
```csharp
public class RtpOptimizationPlotView : PlotView
    {
        public override void OnApplyTemplate()
        {
            base.OnApplyTemplate();
            grid.Children.RemoveAt(grid.Children.Count - 1);
        }
    }
```
3. Add the custom plot view to a user control or so and define the trackercontrol. Note that it is important to set `IsHitTestVisible="True"` for TrackerControl and point to the correct `DataContext` for the Buttons' commands:
```xaml
<UserControl 
    ...
    x:Name="RtOptimizationUserControl">
<views:RtpOptimizationPlotView
      Model="{Binding OptimizationPlotModel, Mode=TwoWay}"
      Controller="{Binding CustomPlotController}">
      <views:RtpOptimizationPlotView.DefaultTrackerTemplate>
          <ControlTemplate>
              <oxy:TrackerControl
                  IsHitTestVisible="True"
                  Position="{Binding Position}" LineExtents="{Binding PlotModel.PlotArea}"
                  BorderThickness="1">
                  <oxy:TrackerControl.Background>
                      <SolidColorBrush Color="LightBlue" />
                  </oxy:TrackerControl.Background>
                  <oxy:TrackerControl.Content>
                      <StackPanel Orientation="Vertical">
                          <TextBlock Text="{Binding}" Margin="7"/>
                          <StackPanel Orientation="Horizontal">
                              <Button 
                                  Command="{Binding DataContext.FirstButtonsCommand, ElementName=RtOptimizationUserControl}">Button 1</Button>
                              <Button
                                  Command="{Binding DataContext.SecondButtonsCommand, ElementName=RtOptimizationUserControl}">Button 2</Button>
                          </StackPanel>
                      </StackPanel>
                  </oxy:TrackerControl.Content>
              </oxy:TrackerControl>
          </ControlTemplate>
      </views:RtpOptimizationPlotView.DefaultTrackerTemplate>
  </views:RtpOptimizationPlotView>
</UserControl>
```

Thanks a lot to @VisualMelon for all your help!

**Reply by VisualMelon** [reply]
Thanks for providing this write-up: will be very useful for other people in future to have it clearly presented in one place like this.
